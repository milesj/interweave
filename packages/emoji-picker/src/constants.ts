/* eslint-disable sort-keys */

import {
	GROUP_KEY_ACTIVITIES,
	GROUP_KEY_ANIMALS_NATURE,
	GROUP_KEY_COMPONENT,
	GROUP_KEY_FLAGS,
	GROUP_KEY_FOOD_DRINK,
	GROUP_KEY_OBJECTS,
	GROUP_KEY_PEOPLE_BODY,
	GROUP_KEY_SMILEYS_EMOTION,
	GROUP_KEY_SYMBOLS,
	GROUP_KEY_TRAVEL_PLACES,
	SKIN_KEY_DARK,
	SKIN_KEY_LIGHT,
	SKIN_KEY_MEDIUM,
	SKIN_KEY_MEDIUM_DARK,
	SKIN_KEY_MEDIUM_LIGHT,
} from 'emojibase';
import { GroupKey, SkinToneKey } from './types';

export {
	GROUP_KEY_ACTIVITIES,
	GROUP_KEY_ANIMALS_NATURE,
	GROUP_KEY_COMPONENT,
	GROUP_KEY_FLAGS,
	GROUP_KEY_FOOD_DRINK,
	GROUP_KEY_OBJECTS,
	GROUP_KEY_PEOPLE_BODY,
	GROUP_KEY_SMILEYS_EMOTION,
	GROUP_KEY_SYMBOLS,
	GROUP_KEY_TRAVEL_PLACES,
	SKIN_KEY_DARK,
	SKIN_KEY_LIGHT,
	SKIN_KEY_MEDIUM,
	SKIN_KEY_MEDIUM_DARK,
	SKIN_KEY_MEDIUM_LIGHT,
};

export const GROUP_KEY_COMMONLY_USED = 'commonly-used';
export const GROUP_KEY_SEARCH_RESULTS = 'search-results';
export const GROUP_KEY_VARIATIONS = 'variations';
export const GROUP_KEY_NONE = 'none';

export const GROUPS: GroupKey[] = [
	GROUP_KEY_SMILEYS_EMOTION,
	GROUP_KEY_PEOPLE_BODY,
	GROUP_KEY_COMPONENT, // Unused but required for order
	GROUP_KEY_ANIMALS_NATURE,
	GROUP_KEY_FOOD_DRINK,
	GROUP_KEY_TRAVEL_PLACES,
	GROUP_KEY_ACTIVITIES,
	GROUP_KEY_OBJECTS,
	GROUP_KEY_SYMBOLS,
	GROUP_KEY_FLAGS,
];

export const GROUP_ICONS: Record<string, string> = {
	[GROUP_KEY_COMMONLY_USED]: 'üïë',
	[GROUP_KEY_SMILEYS_EMOTION]: 'üòÉ',
	[GROUP_KEY_PEOPLE_BODY]: 'üëç',
	[GROUP_KEY_ANIMALS_NATURE]: 'üåø',
	[GROUP_KEY_FOOD_DRINK]: 'üçé',
	[GROUP_KEY_TRAVEL_PLACES]: 'üó∫Ô∏è',
	[GROUP_KEY_ACTIVITIES]: '‚öΩÔ∏è',
	[GROUP_KEY_OBJECTS]: 'üìò',
	[GROUP_KEY_SYMBOLS]: '‚õîÔ∏è',
	[GROUP_KEY_FLAGS]: 'üè¥',
};

export const SKIN_KEY_NONE = 'none';

export const SKIN_TONES: SkinToneKey[] = [
	SKIN_KEY_NONE,
	SKIN_KEY_LIGHT,
	SKIN_KEY_MEDIUM_LIGHT,
	SKIN_KEY_MEDIUM,
	SKIN_KEY_MEDIUM_DARK,
	SKIN_KEY_DARK,
];

export const SKIN_COLORS = {
	[SKIN_KEY_NONE]: '#FFCC22',
	[SKIN_KEY_LIGHT]: '#FADCBC',
	[SKIN_KEY_MEDIUM_LIGHT]: '#E0BB95',
	[SKIN_KEY_MEDIUM]: '#BF8F68',
	[SKIN_KEY_MEDIUM_DARK]: '#9B643D',
	[SKIN_KEY_DARK]: '#5A463A',
};

export const SCROLL_BUFFER = 150;
export const SCROLL_DEBOUNCE = 100;
export const SEARCH_THROTTLE = 300;

export const KEY_COMMONLY_USED = 'interweave/emoji/commonlyUsed';
export const KEY_SKIN_TONE = 'interweave/emoji/skinTone';

export const COMMON_MODE_RECENT = 'recently-used';
export const COMMON_MODE_FREQUENT = 'frequently-used';

export const CONTEXT_CLASSNAMES = {
	picker: 'interweave-picker__picker',
	emoji: 'interweave-picker__emoji',
	emojiActive: 'interweave-picker__emoji--active',
	emojis: 'interweave-picker__emojis',
	emojisList: 'interweave-picker__emojis-list',
	emojisRow: 'interweave-picker__emojis-row',
	emojisHeader: 'interweave-picker__emojis-header',
	emojisHeaderSticky: 'interweave-picker__emojis-header--sticky',
	emojisBody: 'interweave-picker__emojis-body',
	group: 'interweave-picker__group',
	groupActive: 'interweave-picker__group--active',
	groups: 'interweave-picker__groups',
	groupsList: 'interweave-picker__groups-list',
	skinTone: 'interweave-picker__skin-tone',
	skinToneActive: 'interweave-picker__skin-tone--active',
	skinTones: 'interweave-picker__skin-tones',
	skinTonesList: 'interweave-picker__skin-tones-list',
	noPreview: 'interweave-picker__no-preview',
	noResults: 'interweave-picker__no-results',
	preview: 'interweave-picker__preview',
	previewEmoji: 'interweave-picker__preview-emoji',
	previewContent: 'interweave-picker__preview-content',
	previewTitle: 'interweave-picker__preview-title',
	previewSubtitle: 'interweave-picker__preview-subtitle',
	previewShiftMore: 'interweave-picker__preview-more',
	search: 'interweave-picker__search',
	searchInput: 'interweave-picker__search-input',
	clear: 'interweave-picker__clear',
};

export const CONTEXT_MESSAGES = {
	frequentlyUsed: 'Frequently used',
	recentlyUsed: 'Recently used',
	variations: 'Variations',
	none: 'All emojis',
	// Skin tones
	skinNone: 'No skin tone',
	skinLight: 'Light skin tone',
	skinMediumLight: 'Medium-light skin tone',
	skinMedium: 'Medium skin tone',
	skinMediumDark: 'Medium-dark skin tone',
	skinDark: 'Dark skin tone',
	// Miscellaneous
	search: 'Search',
	searchA11y: 'Search for emojis by keyword',
	searchResults: 'Search results',
	noPreview: '',
	noResults: 'No results',
	clearUsed: 'Clear used',
};
