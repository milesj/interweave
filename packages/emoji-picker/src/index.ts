/**
 * @copyright   2016-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Picker } from './Picker';

export * from './constants';
export * from './types';

export default Picker;
