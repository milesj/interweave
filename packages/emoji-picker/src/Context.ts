import React from 'react';
import { EmojiDataManager, LATEST_DATASET_VERSION } from 'interweave-emoji';
import { CONTEXT_CLASSNAMES, CONTEXT_MESSAGES } from './constants';
import { Context as ContextType } from './types';

export const Context = React.createContext<ContextType>({
	classNames: CONTEXT_CLASSNAMES,
	emojiData: EmojiDataManager.getInstance('en'),
	emojiLargeSize: 0,
	emojiPadding: 0,
	emojiPath: '{{hexcode}}',
	emojiSize: 0,
	emojiSource: {
		compact: false,
		locale: 'en',
		version: LATEST_DATASET_VERSION,
	},
	messages: CONTEXT_MESSAGES,
});
