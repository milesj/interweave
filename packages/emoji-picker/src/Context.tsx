import React from 'react';
import { Context } from './types';
import { CONTEXT_CLASSNAMES, CONTEXT_MESSAGES } from './constants';

export default React.createContext<Context>({
  classNames: CONTEXT_CLASSNAMES,
  messages: CONTEXT_MESSAGES,
});
