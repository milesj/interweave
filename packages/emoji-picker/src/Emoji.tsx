import React, { useContext } from 'react';
import { Emoji as EmojiCharacter, CanonicalEmoji } from 'interweave-emoji';
import Context from './Context';

export interface EmojiProps {
  active: boolean;
  emoji: CanonicalEmoji;
  onEnter: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
  onLeave: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
  onSelect: (emoji: CanonicalEmoji, event: React.MouseEvent<HTMLButtonElement>) => void;
}

export default function Emoji({ active, emoji, onEnter, onLeave, onSelect }: EmojiProps) {
  const { classNames, emojiPadding, emojiPath, emojiSize, emojiSource } = useContext(Context);
  const dimension = emojiPadding + emojiPadding + emojiSize;
  const className = [classNames.emoji];

  if (active) {
    className.push(classNames.emojiActive);
  }

  // Handlers
  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    onSelect(emoji, event);
  };

  const handleEnter = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    onEnter(emoji, event);
  };

  const handleLeave = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    onLeave(emoji, event);
  };

  return (
    <button
      key={emoji.hexcode}
      className={className.join(' ')}
      style={{
        height: dimension,
        padding: emojiPadding,
        width: dimension,
      }}
      title={emoji.annotation}
      type="button"
      onClick={handleClick}
      onMouseEnter={handleEnter}
      onMouseLeave={handleLeave}
    >
      <EmojiCharacter
        path={emojiPath}
        size={emojiSize}
        source={emojiSource}
        hexcode={emoji.hexcode}
      />
    </button>
  );
}
