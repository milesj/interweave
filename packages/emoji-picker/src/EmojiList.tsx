import React, { useCallback, useContext, useEffect, useRef, useState } from 'react';
import { FixedSizeList as List, ListOnItemsRenderedProps } from 'react-window';
import chunk from 'lodash/chunk';
import { GROUP_KEY_COMPONENT, GROUP_KEY_NONE } from './constants';
import Context from './Context';
import EmojiListHeader from './EmojiListHeader';
import EmojiListRow, { EmojiListRowProps, VirtualRow } from './EmojiListRow';
import { GroupEmojiMap, GroupKey } from './types';

export interface IndicesMap {
	[key: string]: number;
}

export interface EmojiListProps extends EmojiListRowProps {
	activeGroup: GroupKey;
	columnCount: number;
	columnPadding?: number;
	groupedEmojis: GroupEmojiMap;
	hideGroupHeaders: boolean;
	noResults?: React.ReactNode;
	onScroll: () => void;
	onScrollGroup: (group: GroupKey) => void;
	rowCount: number;
	rowPadding?: number;
	scrollToGroup?: GroupKey | '';
	stickyGroupHeader?: boolean;
}

export default function EmojiList({
	activeGroup,
	columnCount,
	columnPadding = 0,
	groupedEmojis,
	hideGroupHeaders,
	noResults,
	rowCount,
	rowPadding = 0,
	scrollToGroup,
	stickyGroupHeader,
	onScroll,
	onScrollGroup,
	...rowProps
}: EmojiListProps) {
	const { classNames, emojiPadding, emojiSize, messages } = useContext(Context);
	const [rows, setRows] = useState<VirtualRow[]>([]);
	const [indices, setIndices] = useState<IndicesMap>({});
	const ref = useRef<List>(null);
	const size = emojiSize + emojiPadding * 2;
	const rowHeight = size + rowPadding * 2;
	const columnWidth = size + columnPadding * 2;

	// When emojis or virtual list props change,
	// we need to regenerate the list of rows.
	useEffect(() => {
		const virtualRows: VirtualRow[] = [];
		const nextIndices: IndicesMap = {
			'': -1, // Handle empty scroll to's
		};

		Object.keys(groupedEmojis).forEach((group) => {
			nextIndices[group] = virtualRows.length;

			if (group === GROUP_KEY_COMPONENT) {
				return;
			}

			if (!hideGroupHeaders) {
				virtualRows.push(group);
			}

			virtualRows.push(...chunk(groupedEmojis[group].emojis, columnCount));
		});

		setRows(virtualRows);
		setIndices(nextIndices);
	}, [columnCount, groupedEmojis, hideGroupHeaders]);

	// Scroll to the defined group when all data is available
	useEffect(() => {
		if (ref.current && scrollToGroup && indices[scrollToGroup] >= 0) {
			ref.current.scrollToItem(indices[scrollToGroup], 'start');
		}
	}, [scrollToGroup, indices]);

	// Loop over each group section within the scrollable container
	// and determine the active group.
	const handleRendered = useCallback(
		({ visibleStartIndex }: ListOnItemsRenderedProps) => {
			let lastGroup = '';

			Object.keys(indices).some((group) => {
				const index = indices[group];

				// Special case for commonly used and smileys, as they usually both render in the same view
				if (index === 0 && visibleStartIndex === 0) {
					lastGroup = group;

					return true;
				}

				// When we have to sticky headers, we need to change the header on the index right
				// before the next header, otherwise the change will happen too late
				if (stickyGroupHeader && index >= visibleStartIndex + 1) {
					return true;
					// Otherwise, we should update the active group when half way through the list
				} else if (!stickyGroupHeader && index >= visibleStartIndex + rowCount / 2) {
					return true;
				}

				lastGroup = group;

				return false;
			});

			// Only update if a different group
			if (lastGroup && lastGroup !== activeGroup) {
				onScrollGroup(lastGroup as GroupKey);
			}
		},
		[activeGroup, indices, onScrollGroup, rowCount, stickyGroupHeader],
	);

	// If no items to display, just return null
	if (rows.length === 0) {
		return <div className={classNames.noResults}>{noResults || messages.noResults}</div>;
	}

	return (
		<div className={classNames.emojis}>
			<List
				ref={ref}
				className={classNames.emojisList}
				height={rowHeight * rowCount}
				itemCount={rows.length}
				itemData={rows}
				itemSize={rowHeight}
				overscanCount={rowCount / 2}
				width={columnWidth * columnCount}
				onItemsRendered={handleRendered}
				onScroll={onScroll}
			>
				{(props) => <EmojiListRow {...rowProps} {...props} />}
			</List>

			{stickyGroupHeader && activeGroup !== GROUP_KEY_NONE && (
				<EmojiListHeader {...rowProps} sticky group={activeGroup} />
			)}
		</div>
	);
}
