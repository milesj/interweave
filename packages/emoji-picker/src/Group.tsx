import React, { useCallback, useContext } from 'react';
import { Context } from './Context';
import { useGroupMessage } from './hooks/useGroupMessage';
import { CommonMode, GroupKey } from './types';

export interface GroupProps {
	active: boolean;
	children: React.ReactNode;
	commonMode: CommonMode;
	group: GroupKey;
	onSelect: (group: GroupKey, event: React.MouseEvent<HTMLButtonElement>) => void;
}

export function Group({ active, children, commonMode, group, onSelect }: GroupProps) {
	const { classNames } = useContext(Context);
	const className = [classNames.group];
	const title = useGroupMessage(group, commonMode);

	if (active) {
		className.push(classNames.groupActive);
	}

	const handleClick = useCallback(
		(event: React.MouseEvent<HTMLButtonElement>) => {
			event.stopPropagation();
			onSelect(group, event);
		},
		[group, onSelect],
	);

	return (
		<button className={className.join(' ')} title={title} type="button" onClick={handleClick}>
			{children}
		</button>
	);
}
