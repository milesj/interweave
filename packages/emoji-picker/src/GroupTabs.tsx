import React, { useContext } from 'react';
import { CanonicalEmoji } from 'interweave-emoji';
import camelCase from 'lodash/camelCase';
import { GROUP_ICONS, GROUP_KEY_COMMONLY_USED, GROUP_KEY_COMPONENT, GROUPS } from './constants';
import { Context } from './Context';
import { Group } from './Group';
import { CommonMode, GroupKey } from './types';

export interface GroupTabsProps {
	activeGroup: GroupKey;
	commonEmojis: CanonicalEmoji[];
	commonMode: CommonMode;
	icons: Record<string, React.ReactNode>;
	onSelect: (group: GroupKey, event: React.MouseEvent<HTMLButtonElement>) => void;
}

export function GroupTabs({
	activeGroup,
	commonEmojis,
	commonMode,
	icons,
	onSelect,
}: GroupTabsProps) {
	const { classNames } = useContext(Context);
	const groups = GROUPS.filter((group) => group !== GROUP_KEY_COMPONENT);

	if (commonEmojis.length > 0) {
		groups.unshift(GROUP_KEY_COMMONLY_USED);
	}

	return (
		<nav className={classNames.groups}>
			<ul className={classNames.groupsList}>
				{groups.map((group) => (
					<li key={group}>
						<Group
							active={group === activeGroup}
							commonMode={commonMode}
							group={group}
							onSelect={onSelect}
						>
							{icons[group] ?? icons[camelCase(group)] ?? GROUP_ICONS[group]}
						</Group>
					</li>
				))}
			</ul>
		</nav>
	);
}
