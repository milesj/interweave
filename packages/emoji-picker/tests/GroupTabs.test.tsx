import React from 'react';
import { DomElement, render } from 'rut-dom';
import {
	COMMON_MODE_FREQUENT,
	GROUP_KEY_ACTIVITIES,
	GROUP_KEY_ANIMALS_NATURE,
	GROUP_KEY_COMMONLY_USED,
	GROUP_KEY_NONE,
	GROUPS,
} from '../src/constants';
import { Group } from '../src/Group';
import { GroupTabs, GroupTabsProps } from '../src/GroupTabs';
import { CAT_EMOJI, ContextWrapper } from './mocks';

function findGroupByKey(root: DomElement<typeof GroupTabs, {}>, key: string) {
	return root
		.query<typeof Group>((node, fiber) => node.type === 'li' && fiber.key === key)[0]
		.findOne(Group);
}

describe('GroupTabs', () => {
	const props: GroupTabsProps = {
		activeGroup: GROUP_KEY_NONE,
		commonEmojis: [],
		commonMode: COMMON_MODE_FREQUENT,
		icons: {},
		onSelect() {},
	};

	it('renders a list', () => {
		const { root } = render<GroupTabsProps>(<GroupTabs {...props} />);

		expect(root.findOne('nav')).toMatchSnapshot();
	});

	it('renders a node for each group', () => {
		const { root } = render<GroupTabsProps>(<GroupTabs {...props} />);

		expect(root.find(Group)).toHaveLength(GROUPS.length - 1);
	});

	it('adds common mode group if there are common emojis', () => {
		const { root } = render<GroupTabsProps>(<GroupTabs {...props} commonEmojis={[CAT_EMOJI]} />);

		expect(root.find(Group)).toHaveLength(GROUPS.length);
		expect(root.findAt(Group, 0)).toHaveProp('group', GROUP_KEY_COMMONLY_USED);
	});

	it('can customize class name', () => {
		const { root } = render<GroupTabsProps>(<GroupTabs {...props} />, {
			wrapper: (
				<ContextWrapper
					classNames={{
						groups: 'test-groups',
					}}
				/>
			),
		});

		expect(root.findOne('nav')).toHaveProp('className', 'test-groups');
	});

	it('passes `onSelect` to group tabs', () => {
		const { root } = render<GroupTabsProps>(<GroupTabs {...props} />);

		expect(root.findAt(Group, 0)).toHaveProp('onSelect', props.onSelect);
	});

	it('sets active group', () => {
		const { root } = render<GroupTabsProps>(
			<GroupTabs {...props} activeGroup={GROUP_KEY_ACTIVITIES} />,
		);

		expect(findGroupByKey(root, GROUP_KEY_ANIMALS_NATURE)).toHaveProp('active', false);
		expect(findGroupByKey(root, GROUP_KEY_ACTIVITIES)).toHaveProp('active', true);
	});

	it('renders default icon', () => {
		const { root } = render<GroupTabsProps>(<GroupTabs {...props} />);

		expect(findGroupByKey(root, GROUP_KEY_ANIMALS_NATURE)).toContainNode('ðŸŒ¿');
	});

	it('renders icon using kebab case', () => {
		const icon = <b>Animal</b>;
		const { root } = render<GroupTabsProps>(
			<GroupTabs {...props} icons={{ 'animals-nature': icon }} />,
		);

		expect(findGroupByKey(root, GROUP_KEY_ANIMALS_NATURE)).toContainNode(icon);
	});

	it('renders icon using camel case', () => {
		const icon = <b>Animal</b>;
		const { root } = render<GroupTabsProps>(
			<GroupTabs {...props} icons={{ animalsNature: icon }} />,
		);

		expect(findGroupByKey(root, GROUP_KEY_ANIMALS_NATURE)).toContainNode(icon);
	});
});
