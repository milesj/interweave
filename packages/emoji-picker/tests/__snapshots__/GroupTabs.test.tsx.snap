// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupTabs renders a list 1`] = `
<nav className="interweave-picker__groups">
  <ul className="interweave-picker__groups-list">
    <li>
      <Group
        active={false}
        commonMode="frequently-used"
        group="smileys-people"
        onSelect={onSelect()}
      >
        <button
          className="interweave-picker__group"
          title="Smileys & People"
          type="button"
          onClick={handleClick()}
        >
          😃
        </button>
      </Group>
    </li>
    <li>
      <Group
        active={false}
        commonMode="frequently-used"
        group="animals-nature"
        onSelect={onSelect()}
      >
        <button
          className="interweave-picker__group"
          title="Animals & Nature"
          type="button"
          onClick={handleClick()}
        >
          🌿
        </button>
      </Group>
    </li>
    <li>
      <Group
        active={false}
        commonMode="frequently-used"
        group="food-drink"
        onSelect={onSelect()}
      >
        <button
          className="interweave-picker__group"
          title="Food & Drink"
          type="button"
          onClick={handleClick()}
        >
          🍎
        </button>
      </Group>
    </li>
    <li>
      <Group
        active={false}
        commonMode="frequently-used"
        group="travel-places"
        onSelect={onSelect()}
      >
        <button
          className="interweave-picker__group"
          title="Travel & Places"
          type="button"
          onClick={handleClick()}
        >
          🗺️
        </button>
      </Group>
    </li>
    <li>
      <Group
        active={false}
        commonMode="frequently-used"
        group="activities"
        onSelect={onSelect()}
      >
        <button
          className="interweave-picker__group"
          title="Activities"
          type="button"
          onClick={handleClick()}
        >
          ⚽️
        </button>
      </Group>
    </li>
    <li>
      <Group
        active={false}
        commonMode="frequently-used"
        group="objects"
        onSelect={onSelect()}
      >
        <button
          className="interweave-picker__group"
          title="Objects"
          type="button"
          onClick={handleClick()}
        >
          📘
        </button>
      </Group>
    </li>
    <li>
      <Group
        active={false}
        commonMode="frequently-used"
        group="symbols"
        onSelect={onSelect()}
      >
        <button
          className="interweave-picker__group"
          title="Symbols"
          type="button"
          onClick={handleClick()}
        >
          ⛔️
        </button>
      </Group>
    </li>
    <li>
      <Group
        active={false}
        commonMode="frequently-used"
        group="flags"
        onSelect={onSelect()}
      >
        <button
          className="interweave-picker__group"
          title="Flags"
          type="button"
          onClick={handleClick()}
        >
          🏴
        </button>
      </Group>
    </li>
  </ul>
</nav>
`;
